src/main/java/com/github/mmolimar/kafka/connect/fs/FsSourceTask.java

```diff
@@ -74,10 +75,17 @@ public void start(Map<String, String> properties) {
     public List<SourceRecord> poll() {
         while (!stop.get() && policy != null && !policy.hasEnded()) {
             log.trace("Polling for new data...");
+            Function<FileMetadata, Map<String, Object>> makePartitionKey = (FileMetadata metadata) ->
+                    Collections.singletonMap("path", metadata.getPath());
 
-            List<SourceRecord> totalRecords = filesToProcess().map(metadata -> {
+            // Fetch all the offsets upfront to avoid fetching offsets once per file
+            List<FileMetadata> filesToProcess = filesToProcess().collect(Collectors.toList());
```

Comment:
This will only fetch offsets for files up to the `policy.batch_size` configuration.